include/drivers/pinctrl.h

```diff
@@ -0,0 +1,340 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * Public APIs for pin control drivers
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_PINCTRL_H_
+#define ZEPHYR_INCLUDE_DRIVERS_PINCTRL_H_
+
+/**
+ * @brief Pin Controller Interface
+ * @defgroup pinctrl_interface Pin Controller Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <device.h>
+#include <devicetree.h>
+#include <devicetree/pinctrl.h>
+#include <sys/util.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** Pin controller state. */
+enum pinctrl_state {
+	/** Default. */
+	PINCTRL_STATE_DEFAULT,
+	/** Sleep. */
+	PINCTRL_STATE_SLEEP,
+	/** This and higher values refer to custom private states. */
+	PINCTRL_STATE_PRIV_START,
+};
+
+/** Pin control state configuration. */
+struct pinctrl_state_config {
+	/** State identifier. */
+	enum pinctrl_state id;
+	/** Pin configurations. */
+	const pinctrl_pin_t *pins;
+	/** Number of pin configurations. */
+	size_t pins_cnt;
+};
+
+/** Pin controller configuration for a given device. */
+struct pinctrl_config {
+#ifdef CONFIG_PINCTRL_STORE_REG
+	/** Device address. */
+	uintptr_t reg;
```

Comment:
I wonder about making this an opaque pointer instead.

Reply:
considering we store reg, isn't already the correct type?